File: 2023-2024-Over_Under/OverUnder2023-main/programming/pidTuning.py
proportional = 0
derivative = 0
integral = 0
ki = 1.2
kd = 1

def PStraight_dist_head_velo_kp(PStraight_dist_head_velo_kp__dist, PStraight_dist_head_velo_kp__head, PStraight_dist_head_velo_kp__velo, PStraight_dist_head_velo_kp__kp):
    global proportional, derivative, kd, integral, ki
    LeftDrive.set_position(0, DEGREES)
    RightDrive.set_position(0, DEGREES)
    LeftDrive.spin(FORWARD)
    RightDrive.spin(FORWARD)
    if PStraight_dist_head_velo_kp__velo > 0:
        while 30 * PStraight_dist_head_velo_kp__dist > LeftDrive.position(DEGREES):
            derivative = PStraight_dist_head_velo_kp__head - inertial_7.rotation(DEGREES)
            proportional = PStraight_dist_head_velo_kp__kp * (kd * derivative)
            LeftDrive.set_velocity((PStraight_dist_head_velo_kp__velo + proportional), PERCENT)
            RightDrive.set_velocity((PStraight_dist_head_velo_kp__velo - proportional), PERCENT)
            wait(5, MSEC)
    else:
        while LeftDrive.position(DEGREES) > 30 * PStraight_dist_head_velo_kp__dist:
            derivative = PStraight_dist_head_velo_kp__head - inertial_7.rotation(DEGREES)
            proportional = PStraight_dist_head_velo_kp__kp * (kd * derivative)
            LeftDrive.set_velocity((PStraight_dist_head_velo_kp__velo + proportional), PERCENT)
            RightDrive.set_velocity((PStraight_dist_head_velo_kp__velo - proportional), PERCENT)
            wait(5, MSEC)
    LeftDrive.stop()
    RightDrive.stop()

def PTurn_head(PTurn_head__head):
    global proportional, derivative, kd, integral, ki
    integral = 90
    LeftDrive.set_position(0, DEGREES)
    RightDrive.set_position(0, DEGREES)
    LeftDrive.spin(FORWARD)
    RightDrive.spin(FORWARD)
    if inertial_7.rotation(DEGREES) > PTurn_head__head:
        while integral > 2:
            integral = ki * (PStraight_dist_head_velo_kp__head - inertial_7.rotation(DEGREES))
            LeftDrive.set_velocity((0 + integral), PERCENT)
            RightDrive.set_velocity((0 - integral), PERCENT)
            wait(5, MSEC)
    else:
        while integral > 2:
            integral = ki * (PStraight_dist_head_velo_kp__head - inertial_7.rotation(DEGREES))
            LeftDrive.set_velocity((0 + integral), PERCENT)
            RightDrive.set_velocity((0 - integral), PERCENT)
            wait(5, MSEC)
    LeftDrive.stop()
    RightDrive.stop()

def init():
    global proportional, derivative, kd, integral, ki
    kd = 1
    ki = 1.2
    PStraight_dist_head_velo_kp(1, 1, 1, 1)

init()

File: 2023-2024-Over_Under/OverUnder2023-main/programming/pidTuning.cpp
// User defined function
void myblockfunction_PStraight_dist_head_velo_kp(double myblockfunction_PStraight_dist_head_velo_kp__dist, double myblockfunction_PStraight_dist_head_velo_kp__head, double myblockfunction_PStraight_dist_head_velo_kp__velo, double myblockfunction_PStraight_dist_head_velo_kp__kp);
// User defined function
void myblockfunction_PTurn_head(double myblockfunction_PTurn_head__head);

float targetangle, proportional, angle, xpos, zpos;

// User defined function
void myblockfunction_PStraight_dist_head_velo_kp(double myblockfunction_PStraight_dist_head_velo_kp__dist, double myblockfunction_PStraight_dist_head_velo_kp__head, double myblockfunction_PStraight_dist_head_velo_kp__velo, double myblockfunction_PStraight_dist_head_velo_kp__kp) {
  LeftDrive.setPosition(0.0, degrees);
  RightDrive.setPosition(0.0, degrees);
  LeftDrive.spin(forward);
  RightDrive.spin(forward);
  if (myblockfunction_PStraight_dist_head_velo_kp__velo > 0.0) {
    while ((30.0 * myblockfunction_PStraight_dist_head_velo_kp__dist > LeftDrive.position(degrees))) {
      angle = myblockfunction_PStraight_dist_head_velo_kp__head - BrainInertial.rotation();
      proportional = angle * myblockfunction_PStraight_dist_head_velo_kp__kp;
      LeftDrive.setVelocity((myblockfunction_PStraight_dist_head_velo_kp__velo + proportional), percent);
      RightDrive.setVelocity((myblockfunction_PStraight_dist_head_velo_kp__velo - proportional), percent);
    wait(20, msec);
    }
  }
  else {
    while ((LeftDrive.position(degrees) > 30.0 * myblockfunction_PStraight_dist_head_velo_kp__dist)) {
      angle = myblockfunction_PStraight_dist_head_velo_kp__head - BrainInertial.rotation();
      proportional = angle * myblockfunction_PStraight_dist_head_velo_kp__kp;
      LeftDrive.setVelocity((myblockfunction_PStraight_dist_head_velo_kp__velo + proportional), percent);
      RightDrive.setVelocity((myblockfunction_PStraight_dist_head_velo_kp__velo - proportional), percent);
    wait(20, msec);
    }
  }
  LeftDrive.stop();
  RightDrive.stop();
}

// User defined function
void myblockfunction_PTurn_head(double myblockfunction_PTurn_head__head) {
  targetangle = 90.0;
  LeftDrive.spin(forward);
  RightDrive.spin(forward);
  while ((targetangle > 2.0)) {
    if (myblockfunction_PTurn_head__head < BrainInertial.rotation()) {
      targetangle = 1.2 * (BrainInertial.rotation() - myblockfunction_PTurn_head__head);
      LeftDrive.setVelocity((0.0 - targetangle), percent);
      RightDrive.setVelocity((0.0 + targetangle), percent);
    }
    else {
      targetangle = 1.2 * (BrainInertial.rotation() - myblockfunction_PTurn_head__head);
      LeftDrive.setVelocity((0.0 - targetangle), percent);
      RightDrive.setVelocity((0.0 + targetangle), percent);
    }
  wait(20, msec);
  }
  LeftDrive.stop();
  RightDrive.stop();
}

// "when started" hat block
int init() {
  LeftDrive.setStopping(hold);
  RightDrive.setStopping(hold);
  BrainInertial.setRotation(0.0, degrees);
  BrainInertial.calibrate();
  while (BrainInertial.isCalibrating()) { task::sleep(50); }
  wait(1.0, seconds);
  myblockfunction_PStraight_dist_head_velo_kp(10.0, 0.0, 100.0, 2.0);
  myblockfunction_PTurn_head(-90.0);
  return 0;
}


int main() {
  init();
}
